File: examples/python/Ant.py
import clr
clr.AddReference("AntRunner.Interface")
import AntRunner.Interface.Ant

from AntRunner.Interface import *

from MapPosition import MapPosition

class MyAnt(AntRunner.Interface.Ant):
	Name = "Python Ant"

	MapWidth = 0
	MapHeight = 0

	CurrentX = 0
	CurrentY = 0

	MyColor = ItemColor.None

	Map = []

	CurrentMode = 0
	LastAction = AntAction.Wait
	SearchPrimary = 0
	SearchSecondary = 0
	
	def Initialize(self, mapWidth, mapHeight, color, startX, startY):
		self.MapWidth = mapWidth
		self.MapHeight = mapHeight
		self.MyColor = color
		self.CurrentX = startX
		self.CurrentY = startY

		self.CurrentMode = 0
		self.LastAction = AntAction.Wait

		self.Map = [[0 for x in xrange(mapWidth)] for y in xrange(mapHeight)] 
		for y in xrange(0, mapHeight):
			self.Map.append(y)
			for x in xrange(0, mapWidth):
				self.Map[y][x] = MapPosition()
		
		self.Map[startY][startX].Known = True		
		self.SetSearchDirection()
		
		
	def Tick(self, state):
		self.ProcessEcho(state.Response)
		self.ProcessGameEvent(state.Event)

		if self.CurrentMode == 0:
			self.MapMode()
	
	def SetAction(self, a):
		self.Action = a
		self.LastAction = a
	
	def ProcessEcho(self, response):
		if response == None:
			return
		
		if self.LastAction == AntAction.EchoRight:
			for i in xrange(1, response.Distance):
				self.Map[self.CurrentY][self.CurrentX + i].Known = True

			if self.CurrentX + response.Distance < self.MapWidth:
				self.Map[self.CurrentY][self.CurrentX + response.Distance].Known = True
				self.Map[self.CurrentY][self.CurrentX + response.Distance].Item = response.Item
		elif self.LastAction == AntAction.EchoDown:
			for i in xrange(1, response.Distance):
				self.Map[self.CurrentY + i][self.CurrentX ].Known = True

			if self.CurrentY + response.Distance < self.MapWidth:
				self.Map[self.CurrentY + response.Distance][self.CurrentX].Known = True
				self.Map[self.CurrentY + response.Distance][self.CurrentX].Item = response.Item
		elif self.LastAction == AntAction.EchoLeft:
			for i in xrange(1, response.Distance):
				self.Map[self.CurrentY][self.CurrentX - i].Known = True

			if self.CurrentX - response.Distance >= 0:
				self.Map[self.CurrentY][self.CurrentX - response.Distance].Known = True
				self.Map[self.CurrentY][self.CurrentX - response.Distance].Item = response.Item
		elif self.LastAction == AntAction.EchoUp:
			for i in xrange(1, response.Distance):
				self.Map[self.CurrentY - i][self.CurrentX ].Known = True

			if self.CurrentY - response.Distance >= 0:
				self.Map[self.CurrentY - response.Distance][self.CurrentX].Known = True
				self.Map[self.CurrentY - response.Distance][self.CurrentX].Item = response.Item
		else:
			pass #Should never come here

	def CheckFlag(self, e, check):
		return e & check == check

	def ProcessGameEvent(self, e):
		if self.CheckFlag(e, GameEvent.CollisionDamage):
			return
		
		if self.LastAction == AntAction.MoveRight:
				self.CurrentX += 1
		elif self.LastAction == AntAction.MoveDown:
				self.CurrentY += 1
		elif self.LastAction == AntAction.MoveLeft:
				self.CurrentX -= 1
		elif self.LastAction == AntAction.MoveUp:
				self.CurrentY -= 1

	def MapMode(self):
		#Check left
		if self.CurrentX > 0 and not self.Map[self.CurrentY][self.CurrentX - 1].Known:
			self.SetAction(AntAction.EchoLeft)
			return

		#Check right
		if self.CurrentX < self.MapWidth - 1 and not self.Map[self.CurrentY][self.CurrentX + 1].Known:
			self.SetAction(AntAction.EchoRight)
			return

		#Check up
		if self.CurrentY > 0 and not self.Map[self.CurrentY - 1][self.CurrentX].Known:
			self.SetAction(AntAction.EchoUp)
			return

		#Check  down
		if self.CurrentY < self.MapHeight - 1 and not self.Map[self.CurrentY + 1][self.CurrentX].Known:
			self.SetAction(AntAction.EchoDown)
			return

		#All tiles next to us are known, move a direction.
		if self.CanMove(self.SearchPrimary):
			self.SetAction(self.SearchPrimary)
			return

		if self.CanMove(self.SearchSecondary):
			self.SetAction(self.SearchSecondary)
			return

		if self.LastAction != self.SearchPrimary and self.CanMove(self.OppositeDirection(self.SearchPrimary)):
			self.SearchPrimary = self.OppositeDirection(self.SearchPrimary)
			self.SetAction(self.SearchPrimary)
			return

		if self.CanMove(self.OppositeDirection(self.SearchSecondary)):
			self.SearchSecondary = self.OppositeDirection(self.SearchSecondary)
			self.SetAction(self.SearchSecondary)
			return
		
		self.SetSearchDirection()

	def SetSearchDirection(self):
		if self.CurrentX == 0 or self.MapWidth / self.CurrentX < 0.9:
			self.SearchPrimary = AntAction.MoveRight
		else:
			self.SearchPrimary = AntAction.MoveLeft

		if self.CurrentY == 0 or self.MapHeight / self.CurrentY < 0.9:
			self.SearchSecondary = AntAction.MoveDown
		else:
			self.SearchSecondary = AntAction.MoveUp
		
	def CanMove(self, a):
		nextSpace = None
		if a == AntAction.MoveRight:
			if self.CurrentX + 1 == self.MapWidth:
				return False
			nextSpace = self.Map[self.CurrentY][self.CurrentX + 1]
		elif a == AntAction.MoveDown:
			if self.CurrentY + 1 == self.MapHeight:
				return False
			nextSpace = self.Map[self.CurrentY + 1][self.CurrentX]
		elif a == AntAction.MoveLeft:
			if self.CurrentX - 1 == -1:
				return False
			nextSpace = self.Map[self.CurrentY][self.CurrentX - 1]
		elif a == AntAction.MoveUp:
			if self.CurrentY - 1 == -1:
				return False
			nextSpace = self.Map[self.CurrentY - 1][self.CurrentX]
		else:
			return True

		return nextSpace.Known and not self.Blocked(nextSpace.Item)
	
	def Blocked(self, i):
		if i == Item.SteelWall or i == Item.BrickWall or i == Item.Bomb:
			return True
		if i == Item.RedHome:
			return self.MyColor != ItemColor.Red
		if i == Item.BlueHome:
			return self.MyColor != ItemColor.Blue
		if i == Item.GreenHome:
			return self.MyColor != ItemColor.Green
		if i == Item.OrangeHome:
			return self.MyColor != ItemColor.Orange
		if i == Item.PinkHome:
			return self.MyColor != ItemColor.Pink
		if i == Item.YellowHome:
			return self.MyColor != ItemColor.Yellow
		if i == Item.GrayHome:
			return self.MyColor != ItemColor.Gray
		if i == Item.WhiteHome:
			return self.MyColor != ItemColor.White

		return False

	def OppositeDirection(self, a):
		if a == AntAction.MoveRight:
			return AntAction.MoveLeft
		if a == AntAction.MoveDown:
			return AntAction.MoveUp
		if a == AntAction.MoveLeft:
			return AntAction.MoveRight
		if a == AntAction.MoveUp:
			return AntAction.MoveDown	

		if a == AntAction.EchoRight:
			return AntAction.EchoLeft
		if a == AntAction.EchoDown:
			return AntAction.EchoUp
		if a == AntAction.EchoLeft:
			return AntAction.EchoRight
		if a == AntAction.EchoUp:
			return AntAction.EchoDown
			
		if a == AntAction.ShootRight:
			return AntAction.ShootLeft
		if a == AntAction.ShootDown:
			return AntAction.ShootUp
		if a == AntAction.ShootLeft:
			return AntAction.ShootRight
		if a == AntAction.ShootUp:
			return AntAction.ShootDown

		return AntAction.Wait

File: examples/ruby/Ant.rb
include AntRunner::Interface

require_relative 'MapPosition.rb'

class MyAnt < AntRunner::Interface::Ant
	def Name
		"Ruby Ant"
	end

	def Initialize(mapWidth, mapHeight, color, startX, startY)
		@MapWidth = mapWidth
		@MapHeight = mapHeight
		@MyColor = color
		@CurrentX = startX
		@CurrentY = startY

		@CurrentMode = 0
		@LastAction = AntAction::Wait

		@Map = Array.new(mapHeight) { Array.new(mapWidth)}
		y = 0
		until y >= mapHeight
			x = 0
			until x >= mapWidth
				@Map[y][x] = MapPosition.new
				x += 1
			end
			y += 1
		end
		
		@Map[startY][startX].Known = true		
		SetSearchDirection()
	end
		
	def Tick(state)
		ProcessEcho(state.Response)
		ProcessGameEvent(state.Event)

		if @CurrentMode == 0
			MapMode()
		end
	end
	
	def SetAction(a)
		@Action = a
		@LastAction = a
	end
	
	def ProcessEcho(response)
		if response.nil?
			return
		end
		
		case @LastAction
		when AntAction::EchoRight
			i = 1
			until i >= response.Distance
				@Map[@CurrentY][@CurrentX + i].Known = true
				i += 1
			end

			if @CurrentX + response.Distance < @MapWidth
				@Map[@CurrentY][@CurrentX + response.Distance].Known = true
				@Map[@CurrentY][@CurrentX + response.Distance].Item = response.Item
			end
		when AntAction::EchoDown
			i = 1
			until i >= response.Distance
				@Map[@CurrentY + i][@CurrentX ].Known = true
				i += 1
			end

			if @CurrentY + response.Distance < @MapHeight
				@Map[@CurrentY + response.Distance][@CurrentX].Known = true
				@Map[@CurrentY + response.Distance][@CurrentX].Item = response.Item
			end
		when AntAction::EchoLeft
			i = 1
			until i >= response.Distance
				@Map[@CurrentY][@CurrentX - i].Known = true
				i += 1
			end

			if @CurrentX - response.Distance >= 0
				@Map[@CurrentY][@CurrentX - response.Distance].Known = true
				@Map[@CurrentY][@CurrentX - response.Distance].Item = response.Item
			end
		when AntAction::EchoUp
			i = 1
			until i >= response.Distance
				@Map[@CurrentY - i][@CurrentX ].Known = true
				i += 1
			end

			if @CurrentY - response.Distance >= 0
				@Map[@CurrentY - response.Distance][@CurrentX].Known = true
				@Map[@CurrentY - response.Distance][@CurrentX].Item = response.Item
			end
		else
			#Should never come here
		end
	end

	def CheckFlag(e, check)
		return e & check == check
	end

	def ProcessGameEvent(e)
		if CheckFlag(e, GameEvent::CollisionDamage)
			return
		end
		
		case @LastAction
		when AntAction::MoveRight
				@CurrentX += 1
		when AntAction::MoveDown
				@CurrentY += 1
		when AntAction::MoveLeft
				@CurrentX -= 1
		when AntAction::MoveUp
				@CurrentY -= 1
		end
	end

	def MapMode()
		#Check left
		if @CurrentX > 0 && !@Map[@CurrentY][@CurrentX - 1].Known
			SetAction(AntAction::EchoLeft)
			return
		end

		#Check right
		if @CurrentX < @MapWidth - 1 && !@Map[@CurrentY][@CurrentX + 1].Known
			SetAction(AntAction::EchoRight)
			return
		end

		#Check up
		if @CurrentY > 0 && !@Map[@CurrentY - 1][@CurrentX].Known
			SetAction(AntAction::EchoUp)
			return
		end

		#Check  down
		if @CurrentY < @MapHeight - 1 && !@Map[@CurrentY + 1][@CurrentX].Known
			SetAction(AntAction::EchoDown)
			return
		end

		#All tiles next to us are known, move a direction.
		if CanMove(@SearchPrimary)
			SetAction(@SearchPrimary)
			return
		end

		if CanMove(@SearchSecondary)
			SetAction(@SearchSecondary)
			return
		end

		if @LastAction != @SearchPrimary && CanMove(OppositeDirection(@SearchPrimary))
			@SearchPrimary = OppositeDirection(@SearchPrimary)
			SetAction(@SearchPrimary)
			return
		end

		if CanMove(OppositeDirection(@SearchSecondary))
			@SearchSecondary = OppositeDirection(@SearchSecondary)
			SetAction(@SearchSecondary)
			return
		end
		
		SetSearchDirection()
	end

	def SetSearchDirection()
		if @CurrentX == 0 or @MapWidth / @CurrentX < 0.9
			@SearchPrimary = AntAction::MoveRight
		else
			@SearchPrimary = AntAction::MoveLeft
		end

		if @CurrentY == 0 or @MapHeight / @CurrentY < 0.9
			@SearchSecondary = AntAction::MoveDown
		else
			@SearchSecondary = AntAction::MoveUp
		end
	end
		
	def CanMove(a)
		nextSpace = nil
		case a 
		when AntAction::MoveRight
			if @CurrentX + 1 == @MapWidth
				return false
			end
			nextSpace = @Map[@CurrentY][@CurrentX + 1]
		when AntAction::MoveDown
			if @CurrentY + 1 == @MapHeight
				return false
			end
			nextSpace = @Map[@CurrentY + 1][@CurrentX]
		when AntAction::MoveLeft
			if @CurrentX - 1 == -1
				return false
			end
			nextSpace = @Map[@CurrentY][@CurrentX - 1]
		when AntAction::MoveUp
			if @CurrentY - 1 == -1
				return false
			end
			nextSpace = @Map[@CurrentY - 1][@CurrentX]
		else
			return true
		end

		return nextSpace.Known && !Blocked(nextSpace.Item)
	end
	
	def Blocked(i)
		case i
		when Item::SteelWall, Item::BrickWall, Item::Bomb
			return true		
		when Item::RedHome
			return @MyColor != ItemColor::Red
		when Item::BlueHome
			return @MyColor != ItemColor::Blue
		when Item::GreenHome
			return @MyColor != ItemColor::Green
		when Item::OrangeHome
			return @MyColor != ItemColor::Orange
		when Item::PinkHome
			return @MyColor != ItemColor::Pink
		when Item::YellowHome
			return @MyColor != ItemColor::Yellow
		when Item::GrayHome
			return @MyColor != ItemColor::Gray
		when Item::WhiteHome
			return @MyColor != ItemColor::White
		else
			return false
		end
	end

	def OppositeDirection(a)
		case a
		when AntAction::MoveRight
			return AntAction::MoveLeft
		when AntAction::MoveDown
			return AntAction::MoveUp
		when AntAction::MoveLeft
			return AntAction::MoveRight
		when AntAction::MoveUp
			return AntAction::MoveDown
		when AntAction::EchoRight
			return AntAction::EchoLeft
		when AntAction::EchoDown
			return AntAction::EchoUp
		when AntAction::EchoLeft
			return AntAction::EchoRight
		when AntAction::EchoUp
			return AntAction::EchoDown			
		when AntAction::ShootRight
			return AntAction::ShootLeft
		when AntAction::ShootDown
			return AntAction::ShootUp
		when AntAction::ShootLeft
			return AntAction::ShootRight
		when AntAction::ShootUp
			return AntAction::ShootDown
		else
			return AntAction::Wait
		end
	end
end